docs/Testing-GitHub-Actions.md

```diff
@@ -0,0 +1,142 @@
+# Testing GitHub Actions
+
+Action development is not complete without robust tests. Tests also help accept contributions from community members, catch issues early and release actions frequently with high confidence. This document describes best practices to test actions, measure code coverage and configure pipelines to run them with every change.
+
+## Test Types
+
+Tests for actions can be categorized into two broad categories.
+
+**Unit tests** are used to validate the business logic like - string manipulations, for loop, processing platform call responses. Unit tests are not expected to trigger the platform calls like making azure CLI / REST calls. Alternatively, developers can mock such platform calls. These tests can be run as part of build steps without requiring any environment setup. Developers can also use them during the development phase.
+
+Unit tests should focus on good code coverage. Every checkin should add tests to cover newly added code. Target is to keep code coverage at 80% and it will be made part of the PR approval checklist.
+
+**Functional tests**** (** also known as end to end tests ) validates end to end functional scenarios. It is important to have a handful of such tests to ensure functionality end to end. A functional test should be designed to cover the real time scenario. Our recommended way to create functional test is to create workflow and execute action with different input, scenario combinations.
+
+## Running Tests:
+
+Once suites of unit tests and function tests are ready, they will be run in different life-cycles of the development process.
+
+#### During Action Authoring:
+
+- Running Unit test using VS Code / VS IDE ( Details to be added )
+- Running functional test from dev box ( Details to be added )
+
+#### Pull Request
+
+During pull request, we want to ensure action logic is tested using unit tests on all platform matrices and there is required code coverage. We have designed the YAML to ensure these coverages. Templates for actions will be set up to run the unit tests for the master and release branches.
```

Comment:
We should give the link to the Yaml file

Reply:
this is guidance, actual walk through with example is in "Appendix" section